common/src/main/java/com/wynntils/features/ui/BulkBuyFeature.java

```diff
@@ -13,46 +13,161 @@
+import com.wynntils.utils.render.Texture;
 import com.wynntils.utils.wynn.ContainerUtils;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import net.minecraft.ChatFormatting;
+import net.minecraft.client.gui.screens.Screen;
+import net.minecraft.client.gui.screens.inventory.ContainerScreen;
+import net.minecraft.client.resources.language.I18n;
 import net.minecraft.network.chat.Component;
 import net.minecraft.world.inventory.AbstractContainerMenu;
 import net.minecraft.world.inventory.ClickType;
 import net.minecraft.world.item.ItemStack;
 import net.neoforged.bus.api.EventPriority;
 import net.neoforged.bus.api.SubscribeEvent;
+import org.lwjgl.glfw.GLFW;
 
 @ConfigCategory(Category.UI)
 public class BulkBuyFeature extends Feature {
     @Persisted
     public final Config<Integer> bulkBuyAmount = new Config<>(4);
 
+    @Persisted
+    public final Config<BulkBuySpeed> bulkBuySpeed = new Config<>(BulkBuySpeed.BALANCED);
+
+    private static final String SHOP_TITLE_SUFFIX = " Shop";
     // Test in BulkBuyFeature_PRICE_PATTERN
     private static final Pattern PRICE_PATTERN = Pattern.compile("§6 - §(?:c✖|a✔) §f(\\d+)§7²");
     private static final ChatFormatting BULK_BUY_ACTIVE_COLOR = ChatFormatting.GREEN;
     private static final StyledText PRICE_STR = StyledText.fromString("§6Price:");
 
+    private BulkBuyWidget bulkBuyWidget;
+    private int bulkBoughtSlotNumber = -1; // Slot number of the thing we're buying
+    private AbstractContainerMenu bulkBoughtContainer = null; // Shop container we're buying from
+    private int bulkBoughtAmount = 0; // Amount remaining that we need to buy
+    private int bulkBoughtPrice = 0; // Price of a single item
+
+    @SubscribeEvent
+    public void onShopOpened(SetSlotEvent.Post e) {
+        // This event handles the first init of the shop, we cannot use ScreenInitEvent for this as
+        // it will be fired before the shop contents are sent by the server
+
+        // Shop titles are in slot 4, eg. §aScroll Shop
+        // Shops are all size 54 for double chest, sometimes size 41 is sent (no idea what it's for)
+        if (e.getSlot() != 4 || e.getContainer().getContainerSize() != 54) return;
+
+        initBulkBuyWidget(McUtils.mc().screen);
+    }
+
+    @SubscribeEvent
+    public void onContainerOpened(ScreenInitEvent.Post e) {
+        // This event handles any reinit required, eg window resize
+        // It will not make a widget on first init
+        initBulkBuyWidget(e.getScreen());
+    }
+
+    private void initBulkBuyWidget(Screen screen) {
+        // Neither event guarantees that they do not re-fire on the same screen
+        if (bulkBoughtSlotNumber != -1) return;
+
+        if (!(screen instanceof ContainerScreen containerScreen)) return;
+        if (!(containerScreen.getMenu() instanceof AbstractContainerMenu acm)
+                || acm.getItems().size() != 90) return;
```

Comment:
You know the drill, multi-line ifs have brackets. I assume this was spotless.

Reply:
Ah yea